### ISSUE TITLE
Defaults of Rule no-underscore-dangle contradict themselves

### ISSUE THREAD

**Issue by ericbartholet** [eslint/eslint#12684]
**What rule do you want to change?**
no-underscore-dangle

**Does this change cause the rule to produce more or fewer warnings?**
More

**How will the change be implemented? (New option, new default behavior, etc.)?**
New default behavior

**Please provide some example code that this change will affect:**

```js
/* eslint no-underscore-dangle: "error" */

class BaseClass {
    _sayFoo() { // No Error
        console.log("foo");
    }
}
class DerivedClass extends BaseClass {
    callSayFoo() {
        this._sayFoo(); // Error
        super._sayFoo(); // Error
    }
}
```

**What does the rule currently do for this code?**
Allows dangling underscores in method names by default because of the default value of the enforceInMethodNames option

**What will the rule do after it's changed?**
Prevent dangling underscores in method names by default because of the changed default value of the enforceInMethodNames option

**Are you willing to submit a pull request to implement this change?**
Yes

**Reasoning**
I've seen form the original implementation of the `enforceInMethodNames` option via [New rule request: naming conventions of methods](https://github.com/eslint/eslint/issues/7065), that its default value was chosen to make this change backwards compatible, however, the default options contradict each other and that is worse; by default, I'm allowed to create a dangled method name that I can't call.  The default value for the `enforceInMethodNames` option should be `true`.

**Comment by kaicataldo** [comment]
I'm not against changing the default value of `enforceInMethodNames ` in our next major release.

**Comment by kaicataldo** [comment]
I think it would also be nice to be able to not warn on the cases listed above when `enforceInMethodNames` is false, but I don't think we can do it reliably without type information.
